src/functions.jl

```diff
@@ -278,6 +278,67 @@ function Base.copy(f::ScalarQuadraticFunction)
     )
 end
 
+"""
+    ScalarNonlinearFunction{T<:Number}(head::Symbol, args::Vector{Any})
+
+The scalar-valued nonlinear function `head(args...)`, represented as a symbolic
+expression tree, with the call operator `head` and ordered arguments in `args`.
+
+## `head`
+
+The `head::Symbol` must be a scalar operator recognized by the model.
+
+By default, the list of supported uni-variate operators is given by the list:
+
+ * [`Nonlinear.DEFAULT_UNIVARIATE_OPERATORS`](@ref)
+
+and the list of supported multi-variate operators is given by the list:
+
+ * [`Nonlinear.DEFAULT_MULTIVARIATE_OPERATORS`](@ref)
+
+Additional operators can be registered by setting a [`UserDefinedFunction`](@ref)
+attribute.
+
+## `args`
+
+The vector `args` contains the arguments to the nonlinear function. If the
+operator is univariate, it must contain one element. Otherwise, it may contain
+multiple elements. Each element must be one of the folowing:
```

Comment:
I see the allowed elements are only scalar types. I'm wondering about cases like `+` `norm` where the input could be a very long list of scalars - will that be performant? If not would allowing vector type (VV/VAF etc) inputs be OK down the track? I'm also wondering about a case like `f(vec1, vec2)` where you are dealing with a function with two vector arguments (e.g. relative entropy) - I'm not sure how to handle that if we only allow `head(args...)` where args are scalars.

Reply:
The current AD system doesn't support vector-valued arguments. We could potentially relax this restriction in future.

Reply:
Ok, I'm thinking about CSP situations (where AD does not apply), so please make it as easy as possible to relax this restriction in the near future (I should be able to contribute).

Reply:
There's nothing stopping this other than convention (the args are `Any[]`, so they can include vectors). So this should work:
```julia
X = reshape(MOI.VariableIndex.(1:4), 2, 2)
f = ScalarNonlinearFunction(:log, Any[ScalarNonlinearFunction(:det, Any[X])])
```
The hard part is adding it into the AD system.

And I don't know what to do if the function returns a vector argument.

Reply:
Nice. Is it worth adding some simple tests for cases like this?